.github/workflows/My-ci-pipeline.yml

```diff
@@ -28,4 +28,79 @@ jobs:
       run: docker build --target test --tag todo-app:test .
     
     - name: BuildAndTestImage
-      run: docker run --env-file .env.test todo-app:test tests
\ No newline at end of file
+      run: docker run --env-file .env.test todo-app:test tests
+
+  Publish:
+    if: ${{ github.ref == 'refs/heads/main' && github.event_name == 'push' }}
+    #if: ${{ github.ref == 'refs/heads/module8' && github.event_name == 'push' }}
+    name: Publish
+    runs-on: ubuntu-latest
+    needs: [build]
+    steps:
+    - uses: actions/checkout@v2
+    
+    - uses: satackey/action-docker-layer-caching@v0.0.11
+      continue-on-error: true
+
+    - name: publish to docker hub
+      run: echo "Publishing!"
+
+    #- name: Build Image
+    #  run: docker build --tag latest .
+    - name: Build Image
+      run: docker build --target production --tag moudule8 .
+
+    - name: Change Tag
+      run: docker tag moudule8 wishy78/module8:module8
+
+    - name: Login to Docker hub
+      run: docker login --username wishy78 --password "${{ secrets.DOCKER_PASSWORD }}"
+
+    - name: Publish to Docker hub
+      run: docker push wishy78/module8:module8
+  
+
+  cd:
+    if: ${{ github.ref == 'refs/heads/main' && github.event_name == 'push' }}
+    #if: ${{ github.ref == 'refs/heads/module8' && github.event_name == 'push' }}
+    name: Build and Deploy
+    runs-on: ubuntu-latest
+    needs: [Publish]
+    steps:
+    - uses: actions/checkout@v2
+    
+    - uses: satackey/action-docker-layer-caching@v0.0.11
+      continue-on-error: true
+
+    - uses: akhileshns/heroku-deploy@v3.12.12 # This is the action
+      with:
+        heroku_api_key: ${{secrets.HEROKU_API_KEY}}
+        heroku_app_name: "module8jon" #Must be unique in Heroku
+        heroku_email: "wishy78@hotmail.com"
+        usedocker: true
+
+    - name: Docker image list 
+      run: docker image ls
+
+    - name: Add Heroku Tag  
+      run: docker tag registry.heroku.com/module8jon/web registry.heroku.com/module8jon/web
+      #run: docker tag wishy78/module8:module8 registry.heroku.com/module8jon/web
+    
+    - name: Push to heroku
+      run: docker push registry.heroku.com/module8jon/web
+
+    - name: Release
+      run: heroku container:release web -a module8jon
```

Comment:
This looks like you're duplicating your deployment - first the action does it for you then you repeat it all manually - that's fine if you're just doing it to learn but unnecessary otherwise. If you're just using the first action to login, then we can either avoid that by setting an appropriate HEROKU_API_KEY environment variable, or by using [the `justlogin`  flag](https://github.com/marketplace/actions/deploy-to-heroku#just-login) that this action provides